File: ts/src/harness-optimization/leakage-gate.ts
import type { LeakageAudit } from "./leakage.js";

/**
 * Verified/exploratory leakage gate (AC-879).
 *
 * TypeScript half of the AC-879 parity pair: the same three-branch gate lives
 * here and in
 * `autocontext/src/autocontext/harness_optimization/leakage_gate.py`, and the
 * shared fixture
 * (`fixtures/harness-optimization/leakage-cases/leakage-cases.json`) proves
 * both languages reach identical advance / non_promotion_grade decisions.
 *
 * Consumes a LeakageAudit and the run mode. Verified runs fail closed on
 * contaminated or unknown status, or on missing prompt provenance. Exploratory
 * runs always advance but are stamped non-promotion-grade. Caller-gated: this
 * function never reads settings, so default runs are unaffected until a caller
 * invokes it.
 */

export interface LeakageGateDecision {
  advance: boolean;
  non_promotion_grade: boolean;
  rationale: string;
}

export function evaluateLeakageGate(
  audit: LeakageAudit,
  mode: string,
  promptProvenance: string,
): LeakageGateDecision {
  if (mode === "exploratory") {
    return {
      advance: true,
      non_promotion_grade: true,
      rationale: "exploratory override: advancing non-promotion-grade regardless of leakage",
    };
  }
  // verified
  const blockers: string[] = [];
  if (audit.status !== "clean") {
    blockers.push(`leakage status ${audit.status}: ${audit.reasons.join("; ")}`);
  }
  if (promptProvenance.trim() === "") {
    blockers.push("missing prompt provenance");
  }
  if (blockers.length > 0) {
    return {
      advance: false,
      non_promotion_grade: true,
      rationale: "verified run blocked: " + blockers.join("; "),
    };
  }
  return { advance: true, non_promotion_grade: false, rationale: "verified run clean" };
}

File: autocontext/src/autocontext/harness_optimization/leakage_gate.py
"""Verified/exploratory leakage gate (AC-879).

Consumes a LeakageAudit and the run mode. Verified runs fail closed on
contaminated or unknown status, or on missing prompt provenance. Exploratory
runs always advance but are stamped non-promotion-grade. Caller-gated: this
function never reads settings, so default runs are unaffected until a caller
invokes it.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from autocontext.harness_optimization.leakage import LeakageAudit

if TYPE_CHECKING:
    from autocontext.config.settings import AppSettings


def leakage_gate_active_for(settings: AppSettings, scenario_name: str) -> bool:
    """True iff the leakage gate is globally enabled AND the scenario is allowlisted.

    The allowlist is the comma-separated ``harness_leakage_gate_scenarios``
    setting; an empty allowlist means no scenario is active even when the global
    flag is on. This is the sole opt-in decision: callers check it and only run
    the leakage audit + gate when it returns True.
    """

    allowlist = {s.strip() for s in settings.harness_leakage_gate_scenarios.split(",") if s.strip()}
    return settings.harness_leakage_gate_enabled and scenario_name in allowlist


@dataclass(frozen=True, slots=True)
class LeakageGateDecision:
    advance: bool
    non_promotion_grade: bool
    rationale: str


def evaluate_leakage_gate(audit: LeakageAudit, mode: str, prompt_provenance: str) -> LeakageGateDecision:
    if mode == "exploratory":
        return LeakageGateDecision(
            advance=True,
            non_promotion_grade=True,
            rationale="exploratory override: advancing non-promotion-grade regardless of leakage",
        )
    # verified
    blockers: list[str] = []
    if audit.status != "clean":
        blockers.append(f"leakage status {audit.status}: {'; '.join(audit.reasons)}")
    if not prompt_provenance.strip():
        blockers.append("missing prompt provenance")
    if blockers:
        return LeakageGateDecision(
            advance=False,
            non_promotion_grade=True,
            rationale="verified run blocked: " + "; ".join(blockers),
        )
    return LeakageGateDecision(advance=True, non_promotion_grade=False, rationale="verified run clean")
